File: compiler/src/feedback/items/TypeDefinitionFeedback.swift
func registerTypeDefinitionFeedback(into context: FeedbackContext) {
    context.register(
        id: "duplicate-field-definition",
        query: Queries.fact(DuplicateFieldDefinition.self),
        rank: { _, _ in .syntax },
        location: { node, _ in (primary: node, secondary: []) },
    ) { writer, node, _ in
        writer.write(node)
        writer.write(" is already defined in this type.")
        writer.writeBreak()
        writer.write("Try renaming this field.")
    }

    context.register(
        id: "duplicate-variant-definition",
        query: Queries.fact(DuplicateVariantDefinition.self),
        rank: { _, _ in .syntax },
        location: { node, _ in (primary: node, secondary: []) },
    ) { writer, node, _ in
        writer.write(node)
        writer.write(" is already defined in this type.")
        writer.writeBreak()
        writer.write("Try renaming this variant.")
    }
}

File: compiler/feedback/src/items/type_definition_feedback.rs
use crate::{FeedbackCtx, FeedbackRank};
use wipple_queries::fact;
use wipple_syntax::statements::type_definition::{
    DuplicateFieldDefinition, DuplicateVariantDefinition,
};

pub fn register(ctx: &mut FeedbackCtx<'_>) {
    ctx.feedback("duplicate-field-definition")
        .query(fact::<DuplicateFieldDefinition>)
        .rank(|_| FeedbackRank::Syntax)
        .display(|_db, writer, node, _| {
            writer.node(node);
            writer.string(" is already defined in this type.");
            writer.line_break();
            writer.string("Try renaming this field.");
        })
        .register();

    ctx.feedback("duplicate-variant-definition")
        .query(fact::<DuplicateVariantDefinition>)
        .rank(|_| FeedbackRank::Syntax)
        .display(|_db, writer, node, _| {
            writer.node(node);
            writer.string(" is already defined in this type.");
            writer.line_break();
            writer.string("Try renaming this variant.");
        })
        .register();
}
